From 4090659d7181112e9ec59c86020d36a1daf32695 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:06 -0400
Subject: [powerpc] boot: Fix do_div for 64bit wrapper

Message-id: <1410545655-205645-498-git-send-email-dzickus@redhat.com>
Patchwork-id: 94401
O-Subject: [RHEL7 PATCH 497/626] powerpc/boot: Fix do_div for 64bit wrapper
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit f4bce2f784706800efcab6830111df9b75c2f199
Author: Cedric Le Goater <clg@fr.ibm.com>
Date:   Thu Apr 24 09:23:25 2014 +0200

    powerpc/boot: Fix do_div for 64bit wrapper

    When the boot wrapper is compiled in 64bit, there is no need to
    use __div64_32.

    Signed-off-by: Cedric Le Goater <clg@fr.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/boot/stdio.c b/arch/powerpc/boot/stdio.c
index 5b57800..a701261 100644
--- a/arch/powerpc/boot/stdio.c
+++ b/arch/powerpc/boot/stdio.c
@@ -21,6 +21,18 @@ size_t strnlen(const char * s, size_t count)
  return sc - s;
 }
 
+#ifdef __powerpc64__
+
+# define do_div(n, base) ({      \
+ unsigned int __base = (base);     \
+ unsigned int __rem;      \
+ __rem = ((unsigned long long)(n)) % __base;   \
+ (n) = ((unsigned long long)(n)) / __base;   \
+ __rem;        \
+})
+
+#else
+
 extern unsigned int __div64_32(unsigned long long *dividend,
           unsigned int divisor);
 
@@ -39,6 +51,8 @@ extern unsigned int __div64_32(unsigned long long *dividend,
  __rem;        \
  })
 
+#endif /* __powerpc64__ */
+
 static int skip_atoi(const char **s)
 {
  int i, c;
-- 
1.7.1